File: python/MP8_PartD.py
from pyspark import SparkContext
from pyspark.sql.types import StructType
from pyspark.sql.types import StructField
from pyspark.sql.types import StringType, IntegerType
from pyspark.sql import SparkSession
from pyspark.sql.functions import desc

sc = SparkContext()
spark = SparkSession.builder.getOrCreate()

####
# 1. Setup : Write a function to load it in an RDD & DataFrame
####

# RDD API
# Columns:
# 0: word (string), 1: year (int), 2: frequency (int), 3: books (int)


# Spark SQL - DataFrame API
def load_data(filepath):
    # Load the file as an RDD of lines
    rdd = sc.textFile(filepath)
    
    # Parse each line into a tuple:
    # 0: word (string), 1: year (int), 2: frequency (int), 3: books (int)
    rdd_parsed = rdd.map(lambda line: line.split("\t")).map(
        lambda tokens: (tokens[0], int(tokens[1]), int(tokens[2]), int(tokens[3]))
    )
    
    # Define a schema for the DataFrame
    schema = StructType([
        StructField("word", StringType(), True),
        StructField("year", IntegerType(), True),
        StructField("frequency", IntegerType(), True),
        StructField("books", IntegerType(), True)
    ])
    
    # Create a DataFrame with the parsed RDD and schema
    df = spark.createDataFrame(rdd_parsed, schema)
    return rdd_parsed, df

# Load the data from "gbooks" (assumed to be in the current directory)
rdd_data, df_data = load_data("gbooks")
####
#  4. MapReduce : List the top three words that have appeared in the
#  greatest number of years. 
#  Sorting order of the final answer should should be descending by word count,
#  then descending by word.

# Spark SQL

# +-------------+--------+
# |         word|count(1)|
# +-------------+--------+
# |    ATTRIBUTE|      11|
# |approximation|       4|
# |    agast_ADV|       4|
# +-------------+--------+
# only showing top 3 rows

df_data.createOrReplaceTempView("gbooks")

# Using Spark SQL to group by word and count its appearances.
# Sorting order: descending by count and descending by word.
result_df = spark.sql("""
    SELECT word, COUNT(*) as `count(1)` 
    FROM gbooks 
    GROUP BY word 
    ORDER BY `count(1)` DESC, word DESC 
""")
result_df.show(3)

File: java/src/main/java/MP8_PartD.java
import scala.Tuple2;

import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.Dataset;
import static org.apache.spark.sql.functions.*;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.types.StringType;
import org.apache.spark.sql.types.IntegerType;

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.regex.Pattern;
import java.util.Map;
//import java.util.function.Function;

public final class MP8_PartD {

  public static void main(String[] args) throws Exception {
    SparkSession spark = SparkSession
      .builder()
      .appName("MP8")
      .getOrCreate();
    JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());
    SQLContext sqlContext = new SQLContext(sc);
    /*
     * 1. Setup: write a function to load it in an RDD & DataFrame
     */
    
    // RDD API
    // Columns: 0: word (string), 1: year (int), 2: frequency (int), 3: books (int)


    // Spark SQL - DataSet API


    /*
     * 4. MapReduce : List the top three words that have appeared in the
     * greatest number of years. Sorting order of the final answer should should be descending by word count,
     * then descending by word.
     */

    // Dataset/Spark SQL API


    spark.stop();
    sc.stop();
  }
}

/* Sample Output 

+-------------+--------+
|         word|count(1)|
+-------------+--------+
|    ATTRIBUTE|      11|
|approximation|       4|
|    agast_ADV|       4|
+-------------+--------+
only showing top 3 rows

*/
